### PR TITLE
Allow more data into `decode` and add filtering capabilities

### PR DESCRIPTION
**Pull request by tlow22**
Pre-this PR, version 0.3.0 had the capability to return 

`city, country` 

This PR makes updates to: 
1. `download_data` to allow the full headers to the cities data to be retrieved 
2. `read_data` reads in CSV data using `CSV.read` with `DataFrame` sink instead of `CSV.File`. 
    - This allows native `DataFrame` filter operations to be included with the `read_data` function, allowing the reverse geocode to operate on a filtered subsets according to user's wishes
4. `Geocoder` constructor now has keyword argument `filters`
5. depending on header selection, the `decode` function now dispenses results containing other headers as well in addition to `city, country`. This all hinges on the constructor of the `Geocoder` object.

### CODE REVIEW

**Comment by BoZenKhaa** [comment]
Hello @tlow22, what a pleasant surprise to come back to from a vacation, a first PR; thank you!  I didn't think anyone was using this 😄

I don't have time to catch up on my old code and your changes right now, but I will get into it on the weekend. 

Could you let me know if you plan to update the README with the new usage options after you are done adding to this PR?

Thank you for taking the time to do this!

**Comment by tlow22** [comment]
Oh man, I'm just glad you're still around, take your time! Yes more than willing to take time to update the README. 

I did this in a rush for the work I do, so very open to feedback if the direction is too drastic

File: src/geocoder.jl
Comment (BoZenKhaa) [inline src/geocoder.jl:43]: I am not sure what is happening here; why is `type==String` checked twice?

File: src/geocoder.jl
Comment (BoZenKhaa) [inline src/geocoder.jl:58]: Put this line inside the `Tuple` on the following line to avoid two variables named `info` with different meanings in the same function.

**Review by BoZenKhaa — CHANGES_REQUESTED** [review]
I reviewed the changes and have some comments in the files. Additionally, could you look into the following points?

-  the tests are now broken due to the use of custom types in the returned string, e.g., `InlineStrings.String31`. Is that an artifact of using DataFrames? A quick fix would be converting all returned values to regular Julia types, which may also be done inside the package. I am not sure about this; maybe it depends on how you use it.
- Could you provide some small examples and test of filters in the tests? 
- Could you update the README with the new usage?

The errors in the tests got me thinking about the `decode` function. How do you use this package? Would adding a `decode` that takes and returns DataFrame make sense? 

Also, I noticed the old version of the package would not even run with fresh Julia installed on my machine, so thanks for fixing that :-)

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl]: Alrighty addressed below!

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl:82]: Ah this is really a `ismissing` vs. not, and in the case of . 

There's a way to avoid this by having some default behaviors on the `CSV.read` for missing entries. I don't mind turning this into an issue then fixed in another PR so that we don't PR scope increase too much

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl:73]: alrighty, addressed!

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl]: yeap, adjusted this as well, so that the inner constructor is not specified and defaults, and other convenience constructors are now outer constructors

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl]: done!

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl]: done! Had to add `DataStructures.OrderedDict` so that we have a `Dict` that preserves order of these keys

File: .gitignore
Comment (tlow22) [reply .gitignore]: done!

**Comment by tlow22** [comment]
> I reviewed the changes and have some comments in the files. Additionally, could you look into the following points?
> 
> * the tests are now broken due to the use of custom types in the returned string, e.g., `InlineStrings.String31`. Is that an artifact of using DataFrames? A quick fix would be converting all returned values to regular Julia types, which may also be done inside the package. I am not sure about this; maybe it depends on how you use it.
> * Could you provide some small examples and test of filters in the tests?
> * Could you update the README with the new usage?
> 
> The errors in the tests got me thinking about the `decode` function. How do you use this package? Would adding a `decode` that takes and returns DataFrame make sense?
> 
> Also, I noticed the old version of the package would not even run with fresh Julia installed on my machine, so thanks for fixing that :-)

So `CSV.read` is now provided the types, allowing a hard typing to `Strings` for each column instead of the `InlineStrings.String31`. More `CSV`'s fault than `DataFrames`. 

I still have yet to add unit tests and examples, but those will come over the week. The previous unit tests broke because `decode` outputs a `NamedTuple`, the values were similar but the order and names weren't. 

I started modifying the code because I wanted the `decode` to output `admin1` to be eventually translated into US states. I actually like `decode`'s behavior now as it's very straightforward to understand.

File: src/geocoder.jl
Comment (BoZenKhaa) [reply src/geocoder.jl]: Oh yeah, I haven't realized this also determines the order of the fields. Nice!

File: src/geocoder.jl
Comment (BoZenKhaa) [reply src/geocoder.jl:82]: Sure, sorry, I think I had a brain cramp and blacked out the `ismissing` when reading this. However, if you want to modify the CSV flags in another PR, by all means :-)

File: Project.toml
Comment (BoZenKhaa) [inline Project.toml:4]: Do you need to increase the patch version in your fork for your work? Anyway, since you are adding functionality, I think I will tag this v0.4.0 when it's merged into the main.

**Comment by BoZenKhaa** [comment]
> > I reviewed the changes and have some comments in the files. Additionally, could you look into the following points?
> > 
> > * the tests are now broken due to the use of custom types in the returned string, e.g., `InlineStrings.String31`. Is that an artifact of using DataFrames? A quick fix would be converting all returned values to regular Julia types, which may also be done inside the package. I am not sure about this; maybe it depends on how you use it.
> > * Could you provide some small examples and test of filters in the tests?
> > * Could you update the README with the new usage?
> > 
> > The errors in the tests got me thinking about the `decode` function. How do you use this package? Would adding a `decode` that takes and returns DataFrame make sense?
> > Also, I noticed the old version of the package would not even run with fresh Julia installed on my machine, so thanks for fixing that :-)
> 
> So `CSV.read` is now provided the types, allowing a hard typing to `Strings` for each column instead of the `InlineStrings.String31`. More `CSV`'s fault than `DataFrames`.
> 
> I still have yet to add unit tests and examples, but those will come over the week. The previous unit tests broke because `decode` outputs a `NamedTuple`, the values were similar but the order and names weren't.
> 
> I started modifying the code because I wanted the `decode` to output `admin1` to be eventually translated into US states. I actually like `decode`'s behavior now as it's very straightforward to understand.

Ok, I will merge this once you are done then. Thank you for working on this!

**Comment by tlow22** [comment]
> > > I reviewed the changes and have some comments in the files. Additionally, could you look into the following points?
> > > 
> > > * the tests are now broken due to the use of custom types in the returned string, e.g., `InlineStrings.String31`. Is that an artifact of using DataFrames? A quick fix would be converting all returned values to regular Julia types, which may also be done inside the package. I am not sure about this; maybe it depends on how you use it.
> > > * Could you provide some small examples and test of filters in the tests?
> > > * Could you update the README with the new usage?
> > > 
> > > The errors in the tests got me thinking about the `decode` function. How do you use this package? Would adding a `decode` that takes and returns DataFrame make sense?
> > > Also, I noticed the old version of the package would not even run with fresh Julia installed on my machine, so thanks for fixing that :-)
> > 
> > 
> > So `CSV.read` is now provided the types, allowing a hard typing to `Strings` for each column instead of the `InlineStrings.String31`. More `CSV`'s fault than `DataFrames`.
> > I still have yet to add unit tests and examples, but those will come over the week. The previous unit tests broke because `decode` outputs a `NamedTuple`, the values were similar but the order and names weren't.
> > I started modifying the code because I wanted the `decode` to output `admin1` to be eventually translated into US states. I actually like `decode`'s behavior now as it's very straightforward to understand.
> 
> Ok, I will merge this once you are done then. Thank you for working on this!

Alright, README + additional unit tests are in! There's other bonuses as well: 

- Added a small improvement to the `Geocoder.country_codes` field where the key is a `Symbol` instead of `String` 
- The `ismissing` confusion you had before? We don't need it anymore.

File: Project.toml
Comment (tlow22) [reply Project.toml]: So we have our in-house registry, I did that because I wasn't sure if you were still around to maintain the code, and wanted a fork that would track my changes.

Once this is merged, I'm more than happy to default back to this instead of the custom package

File: src/geocoder.jl
Comment (tlow22) [reply src/geocoder.jl]: Ah, so this is more a personal preference where I usually keep the default constructor to maintain maximum development flexibility.

**Comment by BoZenKhaa** [comment]
Thank you, this looks great, I will merge this now!

**Pull request merged**
